pkg/expr/sql/parser_allow.go

```diff
@@ -0,0 +1,136 @@
+package sql
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/dolthub/vitess/go/vt/sqlparser"
+)
+
+// AllowQuery parses the query and checks it against an allow list of allowed SQL nodes
+// and functions.
+func AllowQuery(rawSQL string) (bool, error) {
+	s, err := sqlparser.Parse(rawSQL)
+	if err != nil {
+		return false, fmt.Errorf("error parsing sql: %s", err.Error())
+	}
+
+	walkSubtree := func(node sqlparser.SQLNode) error {
+		err := sqlparser.Walk(func(node sqlparser.SQLNode) (bool, error) {
+			if !allowedNode(node) {
+				if fT, ok := node.(*sqlparser.FuncExpr); ok {
+					return false, fmt.Errorf("blocked function %s - not supported in queries", fT.Name)
+				}
+				return false, fmt.Errorf("blocked node %T - not supported in queries", node)
+			}
+			return true, nil
+		}, node)
+
+		if err != nil {
+			return fmt.Errorf("failed to parse SQL expression: %w", err)
+		}
+
+		return nil
+	}
+
+	if err := walkSubtree(s); err != nil {
+		return false, err
+	}
+
+	return true, nil
+}
+
+// nolint:gocyclo,nakedret
+func allowedNode(node sqlparser.SQLNode) (b bool) {
+	b = true // so don't have to return true in every case but default
+
+	switch v := node.(type) {
+	case *sqlparser.FuncExpr:
+		return allowedFunction(v)
+
+	case *sqlparser.AsOf:
+		return
+
+	case *sqlparser.AliasedExpr, *sqlparser.AliasedTableExpr:
+		return
+
+	case *sqlparser.BinaryExpr:
+		return
+
+	case sqlparser.ColIdent, *sqlparser.ColName, sqlparser.Columns:
```

Comment:
This looks great, nice work.

How did you go about assembling this list, please? And how did you choose to use a pointer for `ColName` here, but not the others?

I'm trying to see if I can contribute: eg. was gonna add `ConvertUsingExpr` because I saw:
```go
// ConvertUsingExpr represents a call to CONVERT(expr USING charset).
```

But I thought I'd ask you how to go about getting it right.

Reply:
I ran queries and copied what the error returned.

Sometimes the underlying Type is not an interface, e.g. it will be a struct or `[][]byte` - in those cases it is not a pointer. When under the node interface is another interface, the pointer is needed -- I think.

Reply:
So if you fill out parser_allow_test with things, you will see the type (and if it is a pointer) in the error returned if it is blocked.

Reply:
I think it depends if the methods on the Type defined in vitess ast.go have pointer receivers on their methods or not - maybe?

For example:
```
// Comments represents a list of comments.
type Comments [][]byte

// Format formats the node.
func (node Comments) Format(buf *TrackedBuffer) {
...
}
```

vs

```
type CommonTableExpr struct {
	*AliasedTableExpr
	Columns Columns
}

func (e *CommonTableExpr) Format(buf *TrackedBuffer) {
...
}
```

Reply:
So I would guess if the above is true, `ConvertUsingExpr` will be a pointer, since the methods are pointer recievers:

```
// ConvertUsingExpr represents a call to CONVERT(expr USING charset).
type ConvertUsingExpr struct {
	Expr Expr
	Type string
}

// Format formats the node.
func (node *ConvertUsingExpr) Format(buf *TrackedBuffer) {
	buf.Myprintf("convert(%v using %s)", node.Expr, node.Type)
}
```